src/main/java/htsjdk/samtools/BAMRecord.java

```diff
@@ -256,6 +284,56 @@ public Cigar getCigar() {
         return super.getCigar();
     }
 
+    private static boolean isSentinelCigar(final Cigar cigar, final int readLength) {
+        // There's an implicit assumption here there readLength == length of read in cigar, unless readLength==0
+        return cigar.getCigarElement(0).getOperator() == CigarOperator.S &&
+                (cigar.getCigarElement(0).getLength() == readLength || readLength == 0) &&
+                cigar.getCigarElement(1).getOperator() == CigarOperator.N;
+    }
+
+    private void extractCigarFromCGAttribute(final Cigar sentinelCigar) {
+        final int[] cigarFromCG = (int[]) getAttribute(CG.name());
+
+        if (cigarFromCG == null) return;
+
+        // place the integer array into a buffer so we can decode it
+        final ByteBuffer byteBuffer = ByteBuffer.allocate(cigarFromCG.length * CIGAR_SIZE_MULTIPLIER)
+                .order(ByteOrder.LITTLE_ENDIAN);
+        byteBuffer.asIntBuffer().put(cigarFromCG);
+
+        // decode cigar
+        final Cigar decodedCigar = BinaryCigarCodec.decode(byteBuffer);
+
+        // Sanity check
+        if (decodedCigar.getCigarElements().size() <= MAX_CIGAR_OPERATORS) {
+            throw new IllegalStateException(String.format(
+                    "Only Cigar with > 65535 operators should be placed in CG tag. Found %d operators.",
+                    decodedCigar.getCigarElements().size()));
+        }
+
+        if (decodedCigar.getReferenceLength() != sentinelCigar.getReferenceLength()) {
+            throw new IllegalStateException(String.format(
+                    "Sentinel cigar and CG cigar should have the same reference length. Found %d and %d.",
+                    sentinelCigar.getReferenceLength(),
+                    decodedCigar.getReferenceLength()));
+        }
+
+        if (decodedCigar.getReadLength() != sentinelCigar.getReadLength()) {
+            throw new IllegalStateException(String.format(
+                    "Sentinel cigar and CG cigar should have the same read length. Found %d and %d.",
+                    sentinelCigar.getReadLength(),
+                    decodedCigar.getReadLength()));
+        }
+
+        setCigar(decodedCigar);
```

Comment:
Shouldn't this be a call to `initializeCigar` instead of `setCigar`? This will reset the indexing bin to null.

Reply:
yes.....though now I'm annoyed that none of the tests caught this....what is bin used for ?

Reply:
I think the tests explicitly don't test this.

Reply:
I'm honestly not certain how the bin is used.  I assume it's used in indexing?